## Домашнее задание

1. обучить рукописную сверточную нейронную сеть (с падением размера ядра свертки и последовательностью блоков свертка-пулинг (conv-pool)-(conv-pool)-...) на датасете cifar-10
2. оценить рост точности при увеличении ширины сети (больше фильтров)
3. оценить рост точности при увеличении глубины сети (больше слоев)
4. *сравнить с точностью полносвязной сети для этой выборки

# Задание 1

обучить рукописную сверточную нейронную сеть (с падением размера ядра свертки и последовательностью блоков свертка-пулинг (conv-pool)-(conv-pool)-...) на датасете cifar-10



In [1]:
from tensorflow.keras.datasets import cifar10
from tensorflow.keras.preprocessing.image import ImageDataGenerator
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout, Activation, Flatten
from tensorflow.keras.layers import Conv2D, MaxPooling2D
import os

In [2]:
classes = ['самолет', 'автомобиль', 'птица', 'кот', 'олень', 'собака', 'лягушка', 'лошадь', 'корабль', 'грузовик']

In [68]:
# установка параметров нейросети
batch_size = 512
num_classes = 10
epochs = 5
data_augmentation = False
num_predictions = 20
save_dir = os.path.join(os.getcwd(), 'saved_models')
model_name = 'keras_cifar10_trained_model.h5'

# разделение тренировочной и тестовой выборки
(X_train, y_train), (X_test, y_test) = cifar10.load_data()
print('X_train shape:', X_train.shape)
print(X_train.shape[0], 'тренировочные примеры')
print(X_test.shape[0], 'тестовые примеры')

# преобразование матрицы чисел 0-9 в бинарную матрицу чисел 0-1
y_train = to_categorical(y_train, num_classes)
y_test = to_categorical(y_test, num_classes)


X_train = X_train / 255.0
X_test = X_test / 255.0

X_train shape: (50000, 32, 32, 3)
50000 тренировочные примеры
10000 тестовые примеры


строим архитектуру модели

In [22]:
model_1 = Sequential([
    Conv2D(16, (5, 5), padding='same', input_shape=X_train.shape[1:], activation='relu'),
    MaxPooling2D(pool_size=(2, 2)),
    Conv2D(32, (3, 3), padding='same', input_shape=X_train.shape[1:], activation='relu'),
    MaxPooling2D(pool_size=(2, 2)),
    Flatten(),
    Dense(256, activation='relu', input_shape=(13,)),
    Dense(64, activation='linear'),
    Dense(num_classes, activation='softmax'),
])

model_1.compile(optimizer='adam',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

model_1.summary()

Model: "sequential_7"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d_18 (Conv2D)          (None, 32, 32, 16)        1216      
                                                                 
 max_pooling2d_18 (MaxPoolin  (None, 16, 16, 16)       0         
 g2D)                                                            
                                                                 
 conv2d_19 (Conv2D)          (None, 16, 16, 32)        4640      
                                                                 
 max_pooling2d_19 (MaxPoolin  (None, 8, 8, 32)         0         
 g2D)                                                            
                                                                 
 flatten_6 (Flatten)         (None, 2048)              0         
                                                                 
 dense_21 (Dense)            (None, 256)              

In [23]:
model_1.fit(X_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          validation_data=(X_test, y_test),
          shuffle=True)

Epoch 1/5
98/98 [==============================] - 56s 568ms/step - loss: 1.8218 - accuracy: 0.3435 - val_loss: 1.5356 - val_accuracy: 0.4534
Epoch 2/5
98/98 [==============================] - 56s 567ms/step - loss: 1.3933 - accuracy: 0.5024 - val_loss: 1.3278 - val_accuracy: 0.5214
Epoch 3/5
98/98 [==============================] - 54s 554ms/step - loss: 1.2522 - accuracy: 0.5564 - val_loss: 1.2189 - val_accuracy: 0.5656
Epoch 4/5
98/98 [==============================] - 55s 558ms/step - loss: 1.1545 - accuracy: 0.5932 - val_loss: 1.1535 - val_accuracy: 0.5941
Epoch 5/5
98/98 [==============================] - 55s 558ms/step - loss: 1.0702 - accuracy: 0.6244 - val_loss: 1.0789 - val_accuracy: 0.6169


# Задание 2
оценить рост точности при увеличении ширины сети (больше фильтров)

Увеличим количество каналов на первом конволюционном слое с 32 до 64, на втором conv слое с 16 до 128

In [24]:
model_2 = Sequential([
    Conv2D(64, (5, 5), padding='same', input_shape=X_train.shape[1:], activation='relu'),
    MaxPooling2D(pool_size=(2, 2)),
    Conv2D(128, (3, 3), padding='same', input_shape=X_train.shape[1:], activation='relu'),
    MaxPooling2D(pool_size=(2, 2)),
    Flatten(),
    Dense(256, activation='relu', input_shape=(13,)),
    Dense(64, activation='linear'),
    Dense(num_classes, activation='softmax'),
])

model_2.compile(optimizer='adam',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

model_2.summary()

Model: "sequential_8"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d_20 (Conv2D)          (None, 32, 32, 64)        4864      
                                                                 
 max_pooling2d_20 (MaxPoolin  (None, 16, 16, 64)       0         
 g2D)                                                            
                                                                 
 conv2d_21 (Conv2D)          (None, 16, 16, 128)       73856     
                                                                 
 max_pooling2d_21 (MaxPoolin  (None, 8, 8, 128)        0         
 g2D)                                                            
                                                                 
 flatten_7 (Flatten)         (None, 8192)              0         
                                                                 
 dense_24 (Dense)            (None, 256)              

In [25]:
model_2.fit(X_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          validation_data=(X_test, y_test),
          shuffle=True)

Epoch 1/5
98/98 [==============================] - 218s 2s/step - loss: 1.7289 - accuracy: 0.3785 - val_loss: 1.4695 - val_accuracy: 0.4673
Epoch 2/5
98/98 [==============================] - 219s 2s/step - loss: 1.2566 - accuracy: 0.5538 - val_loss: 1.1524 - val_accuracy: 0.5904
Epoch 3/5
98/98 [==============================] - 219s 2s/step - loss: 1.0726 - accuracy: 0.6221 - val_loss: 1.0386 - val_accuracy: 0.6336
Epoch 4/5
98/98 [==============================] - 216s 2s/step - loss: 0.9408 - accuracy: 0.6681 - val_loss: 1.0004 - val_accuracy: 0.6547
Epoch 5/5
98/98 [==============================] - 217s 2s/step - loss: 0.8521 - accuracy: 0.7027 - val_loss: 0.9208 - val_accuracy: 0.6828


#Вывод:

Благодаря увеличению количества каналов, метрика выросла: на тесте с 0.6169 до 0.6828

# Задание 3

оценить рост точности при увеличении глубины сети (больше слоев)

добавим дополнительные слои конволюции и пулинга в сеть

In [82]:
model_3 = Sequential([
    Conv2D(16, (5, 5), padding='same', input_shape=X_train.shape[1:], activation='relu'),
    MaxPooling2D(pool_size=(2, 2)),
    Conv2D(32, (3, 3), padding='same', input_shape=X_train.shape[1:], activation='relu'),
    MaxPooling2D(pool_size=(2, 2)),

    #ниже дополнительные слои конволюции и пулинга
    Conv2D(64, (3, 3), padding='same', input_shape=X_train.shape[1:], activation='relu'),
    MaxPooling2D(pool_size=(2, 2)),
    Conv2D(128, (3, 3), padding='same', input_shape=X_train.shape[1:], activation='relu'),
    MaxPooling2D(pool_size=(2, 2)),
    
    Flatten(),
    Dense(256, activation='relu', input_shape=(13,)),
    Dense(64, activation='linear'),
    Dense(num_classes, activation='softmax'),
])

model_3.compile(optimizer='adam',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

model_3.summary()

Model: "sequential_14"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d_26 (Conv2D)          (None, 32, 32, 16)        1216      
                                                                 
 max_pooling2d_26 (MaxPoolin  (None, 16, 16, 16)       0         
 g2D)                                                            
                                                                 
 conv2d_27 (Conv2D)          (None, 16, 16, 32)        4640      
                                                                 
 max_pooling2d_27 (MaxPoolin  (None, 8, 8, 32)         0         
 g2D)                                                            
                                                                 
 conv2d_28 (Conv2D)          (None, 8, 8, 64)          18496     
                                                                 
 max_pooling2d_28 (MaxPoolin  (None, 4, 4, 64)       

In [19]:
model_3.fit(X_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          validation_data=(X_test, y_test),
          shuffle=True)

Epoch 1/5
98/98 [==============================] - 73s 740ms/step - loss: 1.8161 - accuracy: 0.3276 - val_loss: 1.5444 - val_accuracy: 0.4310
Epoch 2/5
98/98 [==============================] - 73s 745ms/step - loss: 1.4516 - accuracy: 0.4716 - val_loss: 1.3369 - val_accuracy: 0.5207
Epoch 3/5
98/98 [==============================] - 73s 746ms/step - loss: 1.3092 - accuracy: 0.5291 - val_loss: 1.2570 - val_accuracy: 0.5492
Epoch 4/5
98/98 [==============================] - 73s 742ms/step - loss: 1.1977 - accuracy: 0.5714 - val_loss: 1.2366 - val_accuracy: 0.5580
Epoch 5/5
98/98 [==============================] - 73s 748ms/step - loss: 1.1232 - accuracy: 0.6019 - val_loss: 1.1116 - val_accuracy: 0.6120


# Вывод

увеличение глубины сети не привело к улучшению метрик качества

# Дополнительно

для себя попробовал построить модель с использованием готовой архитектуры ResNet50

In [26]:
from keras.preprocessing.image import ImageDataGenerator
from keras.layers import Dense, Flatten, GlobalAvgPool2D, Dropout, BatchNormalization
from keras.models import Model
from keras.callbacks import ModelCheckpoint

from keras.applications.resnet import ResNet50, preprocess_input as resnet_preproc
from keras.applications.vgg16 import VGG16, preprocess_input as vgg_preproc

import matplotlib.pyplot as plt

In [74]:
# разделение тренировочной и тестовой выборки
(X_train, y_train), (X_test, y_test) = cifar10.load_data()
print('X_train shape:', X_train.shape)
print(X_train.shape[0], 'тренировочные примеры')
print(X_test.shape[0], 'тестовые примеры')

# преобразование матрицы чисел 0-9 в бинарную матрицу чисел 0-1
y_train = to_categorical(y_train, num_classes)
y_test = to_categorical(y_test, num_classes)

X_train shape: (50000, 32, 32, 3)
50000 тренировочные примеры
10000 тестовые примеры


In [75]:
# необходимый препроцессинг данных для ResNet
X_train = resnet_preproc(X_train)
X_test = resnet_preproc(X_test)

In [84]:
input_shape = (32, 32)

В качестве основы сети используем предобученную сеть ResNet50, убираем только последний полносвязный слой

In [76]:
base_model = ResNet50(include_top=False, weights='imagenet', input_shape=(input_shape[0], input_shape[1], 3))
base_model.trainable = False
base_model.summary()

Model: "resnet50"
__________________________________________________________________________________________________
 Layer (type)                   Output Shape         Param #     Connected to                     
 input_6 (InputLayer)           [(None, 32, 32, 3)]  0           []                               
                                                                                                  
 conv1_pad (ZeroPadding2D)      (None, 38, 38, 3)    0           ['input_6[0][0]']                
                                                                                                  
 conv1_conv (Conv2D)            (None, 16, 16, 64)   9472        ['conv1_pad[0][0]']              
                                                                                                  
 conv1_bn (BatchNormalization)  (None, 16, 16, 64)   256         ['conv1_conv[0][0]']             
                                                                                           

добавляем к resnet следующие слои и дообучаемся только на этих слоях:

In [77]:
x = GlobalAvgPool2D()(base_model.output)
x = Dropout(0.1)(x)
x = BatchNormalization()(x)
x = Dense(512, activation='relu')(x)
x = Dense(num_classes, activation='softmax')(x)

model_4 = Model(inputs=base_model.input, outputs=x)
model_4.summary()

Model: "model_5"
__________________________________________________________________________________________________
 Layer (type)                   Output Shape         Param #     Connected to                     
 input_6 (InputLayer)           [(None, 32, 32, 3)]  0           []                               
                                                                                                  
 conv1_pad (ZeroPadding2D)      (None, 38, 38, 3)    0           ['input_6[0][0]']                
                                                                                                  
 conv1_conv (Conv2D)            (None, 16, 16, 64)   9472        ['conv1_pad[0][0]']              
                                                                                                  
 conv1_bn (BatchNormalization)  (None, 16, 16, 64)   256         ['conv1_conv[0][0]']             
                                                                                            

In [78]:
model_4.compile(
    optimizer='adam',
    loss='categorical_crossentropy',
    metrics=['accuracy']
)

In [80]:
model_4.fit(X_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          validation_data=(X_test, y_test),
          shuffle=True)

Epoch 1/5
98/98 [==============================] - 206s 2s/step - loss: 1.2019 - accuracy: 0.6031 - val_loss: 1.0188 - val_accuracy: 0.6496
Epoch 2/5
98/98 [==============================] - 201s 2s/step - loss: 0.8197 - accuracy: 0.7130 - val_loss: 0.9608 - val_accuracy: 0.6721
Epoch 3/5
98/98 [==============================] - 202s 2s/step - loss: 0.6893 - accuracy: 0.7574 - val_loss: 0.9593 - val_accuracy: 0.6722
Epoch 4/5
98/98 [==============================] - 200s 2s/step - loss: 0.5977 - accuracy: 0.7901 - val_loss: 0.9722 - val_accuracy: 0.6777
Epoch 5/5
98/98 [==============================] - 199s 2s/step - loss: 0.5213 - accuracy: 0.8183 - val_loss: 0.9893 - val_accuracy: 0.6721


# Вывод

модель очень сильно переобучается. На трейне метрике высокая, но на тесте она практически не растет.

Юля, напишите пжл в комментарии с чем это связано. Я что-то не так сделал при построении модели?

# Задание 4

*сравнить с точностью полносвязной сети для этой выборки

Строим полносвязную сеть

In [52]:
model_full = Sequential([
    Flatten(input_shape=(32,32,3)),                  
    Dense(256, activation='relu', ),
    Dense(128, activation='linear'),
    Dense(64, activation='relu'),
    Dense(32, activation='linear'),
    Dense(10, activation='softmax')
])

model_full.compile(
    optimizer='adam',
    loss='categorical_crossentropy',
    metrics=['accuracy']
)
model_full.summary()

Model: "sequential_12"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 flatten_11 (Flatten)        (None, 3072)              0         
                                                                 
 dense_50 (Dense)            (None, 256)               786688    
                                                                 
 dense_51 (Dense)            (None, 128)               32896     
                                                                 
 dense_52 (Dense)            (None, 64)                8256      
                                                                 
 dense_53 (Dense)            (None, 32)                2080      
                                                                 
 dense_54 (Dense)            (None, 10)                330       
                                                                 
Total params: 830,250
Trainable params: 830,250
Non-t

In [53]:
model_full.fit(X_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          validation_data=(X_test, y_test),
          shuffle=True)

Epoch 1/5
98/98 [==============================] - 5s 50ms/step - loss: 2.1032 - accuracy: 0.2365 - val_loss: 1.8685 - val_accuracy: 0.3394
Epoch 2/5
98/98 [==============================] - 5s 47ms/step - loss: 1.8245 - accuracy: 0.3475 - val_loss: 1.7759 - val_accuracy: 0.3698
Epoch 3/5
98/98 [==============================] - 5s 47ms/step - loss: 1.7401 - accuracy: 0.3792 - val_loss: 1.7153 - val_accuracy: 0.3814
Epoch 4/5
98/98 [==============================] - 5s 47ms/step - loss: 1.6722 - accuracy: 0.4074 - val_loss: 1.6445 - val_accuracy: 0.4093
Epoch 5/5
98/98 [==============================] - 5s 47ms/step - loss: 1.6252 - accuracy: 0.4203 - val_loss: 1.6267 - val_accuracy: 0.4163


# Вывод:

полносвязная сеть справляетя с задачей гораздо хуже конволюционной